mlir/test/lib/Dialect/Test/TestTypes.cpp

```diff
@@ -525,7 +529,7 @@ void TestRecursiveAliasType::print(AsmPrinter &printer) const {
       printer.tryStartCyclicPrint(*this);
```

Comment:
Would it make sense for `hasFutureAlias` to not be exposed to users at all, but rather be used in the implementation of `tryStartCyclicPrint`? That method is already meant to instruct the user's print implementation whether to print a forward declaration or not. The `hasFutureAlias` condition to me seems like it should be an improvement of that.
If there is a risk of users not wanting this behaviour we could just add a boolean flag disabling it to `tryStartCyclicPrint`.

Not exposing `hasFutureAlias` would also help keep the parse more context insensitive.

Reply:
Yes, I agree here. I don't think we should be exposing more concepts to the user, especially for things like aliases (users shouldn't be thinking about aliases at all, just printing their attribute/type). I would expect that the internals of the printer could be made smarter when printing aliases to make things more compact when they can be.

Reply:
That's a good point. The only concern I have is that the existing `tryStartCyclicPrint` implementation is used to signal that you *must* print a forward declaration, whereas this new check signals that you *may* print a forward declaration. E.g. `TestRecursiveAliasType::parse` has this logic currently:

```cpp
  // If this type already has been parsed above in the stack, expect just the
  // name.
  if (failed(cyclicParse)) {                                                                                                                                                                                                                                                                                                                                                
    if (failed(parser.parseGreater()))
      return Type();
    return rec;
  }
```

With the current implementation, both of the following are valid:

```
# Spell out the mutually recursive type fully
!t1 = !test.test_rec_alias<t1, !test.test_rec_alias<t2, !test.test_rec_alias<t1>>>
!t2 = !test.test_rec_alias<t2, !t1>

# Forward-declare the mutually recursive type
!t1 = !test.test_rec_alias<t1, !test.test_rec_alias<t2>>
!t2 = !test.test_rec_alias<t2, !t1>
```

If we change `tryStartCyclicPrint`, the first one would fail to parse. We could make the new behavior opt-in via a flag, which preserves behavior initially but still doesn't let you distinguish the two cases when you opt-in. We could just accept the new behavior, which I'm warming up to because it limits the number of redundant type definitions which you have to check for consistency. Or we could have a higher-level API that still lets you tell the two cases apart if a dialect needs it. (ClangIR currently prints a self-reference differently from any other incomplete type, but that might be a historical artifact and not an intentional design decision.) I'm fine with any of those options, and I'd appreciate your opinions on them.

Reply:
> used to signal that you must print a forward declaration, whereas this new check signals that you may print a forward declaration

I think I'm missing the distinction here. I don't see why we care to differentiate these two, why would you not print the elided form if you could? My expectation for the parser/printer API for recursive things would be: If the printer tells you that it's okay to elide, you do. In the parser you just check whether you're parsing the elided or non-elided form. As an aside, for this stuff I'd err on changing the contract of the current cyclic printing/parsing logic, it has ~4 uses in tree (2 of which are in tests), I'd hope we can just continue to evolve the recursive support (and not add more constructs).

Reply:
Yeah, you're right. The current parser implementations (both in the test dialect and in ClangIR) were making that distinction but it's not necessary, plus I realized my change would only be affecting printing and not parsing anyway. I'll simplify this accordingly.